azurerm/internal/services/springcloud/spring_cloud_config_server_resource_test.go

```diff
@@ -0,0 +1,207 @@
+	id, err := parse.SpringCloudServiceID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.AppPlatform.ConfigServersClient.Get(ctx, id.ResourceGroup, id.SpringName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read Spring Cloud Service %q (Resource Group %q): %+v", id.SpringName, id.ResourceGroup, err)
+	}
+
+	return utils.Bool(resp.Properties != nil && resp.Properties.ConfigServer != nil && resp.Properties.ConfigServer.GitProperty != nil), nil
+}
+
+func (r SpringCloudConfigServerResource) basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_spring_cloud_config_server" "test" {
+  spring_cloud_service_id = azurerm_spring_cloud_service.test.id
+  uri                     = "https://github.com/Azure-Samples/piggymetrics-config"
+}
+`, r.template(data))
+}
+
+func (r SpringCloudConfigServerResource) requiresImport(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_spring_cloud_config_server" "import" {
+  spring_cloud_service_id = azurerm_spring_cloud_config_server.test.spring_cloud_service_id
+  uri                     = azurerm_spring_cloud_config_server.test.uri
+}
+`, r.basic(data))
+}
+
+func (r SpringCloudConfigServerResource) complete(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_spring_cloud_config_server" "test" {
+  spring_cloud_service_id = azurerm_spring_cloud_service.test.id
+  uri                     = "git@bitbucket.org:Azure-Samples/piggymetrics.git"
+  label                   = "config"
+  search_paths            = ["dir1", "dir4"]
+
+  ssh_auth {
+    private_key                      = file("testdata/private_key")
+    host_key                         = file("testdata/host_key")
+    host_key_algorithm               = "ssh-rsa"
+    strict_host_key_checking_enabled = false
+  }
+
+  repository {
+    name         = "repo1"
+    uri          = "https://github.com/Azure-Samples/piggymetrics"
+    label        = "config"
+    search_paths = ["dir1", "dir2"]
+    http_basic_auth {
+      username = "username"
+      password = "password"
+    }
+  }
+
+  repository {
+    name         = "repo2"
+    uri          = "https://github.com/Azure-Samples/piggymetrics"
+    label        = "config"
+    search_paths = ["dir1", "dir2"]
+  }
+}
+`, r.template(data))
+}
+
+func (SpringCloudConfigServerResource) template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-spring-%d"
+  location = "%s"
+}
+
+resource "azurerm_spring_cloud_service" "test" {
+  name                = "acctest-sc-%d"
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+
+  lifecycle {
+    ignore_changes = [
+      config_server_git_setting,
+    ]
+  }
```

Comment:
This suggests that we still have a breaking change here.  Using a new meta resource shouldn't break the parent, which it appears this does in all cases. I'd recommend taking a look at this config block (`config_server_git_setting` that is) again on the service resource to see if we can handle the deprecation, and the "external" change without this.

Reply:
@jackofallops according to former experience, when split a field into a standalone resource, we would make that field Optional + Computed.  Then encountered a bug that it could not be set to empty, we would set the configMode to `schema.SchemaConfigModeAttr`. There are some examples like this in the codebase. But Tom suggest this will also have a break change when using `dynamic` keyword, and says just left it uncomputed and use ignore_changes. please refer to this link: https://hashicorp.slack.com/archives/G012LHU9XDJ/p1614842151096000

from the api side, I am sorry to say I could not find a way to avoid it

Reply:
@njuCZ ultimately this is a breaking change either way - what use-case is this PR looking to solve exactly?

Reply:
I'm the PM from service team and the use-case is as below in brief:

1. Our enterprise customers use Terraform in both their infra team and apps team.
2. Typically, Infra team mange all the service level configurations like vnet using azurerm_spring_cloud_service, and apps team manage app level configurations using azurerm_spring_cloud_app.
3. Config server is on service level,  but it contains remote settings for all apps so in many cases apps team need to manage configurations here.
4. As infra team and apps team have separate TF repo and pipelines, it's extremely difficult for apps team to manage config server with current service + app resources. A separate resource for config server will perfectly solve this.

Our top customers requested for this change and they are waiting for the fix. Thanks a lot if you could help make that happen.

Reply:
Hi @yucwan - As I understand it, the base config server configuration is common for all apps deployed on the service and _should_ be configured with it. If any infra property is updated, as the whole thing is ForceNew, it will certainly break any deployed application(s) without a co-ordinated effort to apply both team's Terraform runs in the correct order. In the use case you describe, it sounds like the App team should be deploying the `azurerm_spring_cloud_service` using the infrastructure values provided by the infra team, or from data sources, e.g. `location`, `network.*` ? WDYT?

Reply:
@jackofallops "config server configuration is common for all apps" True and that's why we put config server on the service level in our resource model. However the solution does not work for our customers. Infra team take cares of fundamentals like number of service instance, pricing tier and network connectivity/security, thus infra team must control/monitor the `azurerm_spring_cloud_service`. Asking customers changing their workflow for this does not sounds reasonable.

Reply:
@yucwan

Whilst other tooling such as PowerShell handles these delta/patch deployments by updating the field, they don't track changes to the values of these fields - Terraform by contrast tracks the values of these fields in order to determine when there's a diff between the value in Azure and the local value.

Since we already expose the ability to configure the settings at the top level (the service) unfortunately there's no way for us to split these out without making the use-case for the 80% of users harder - or without it being a breaking change.

Based on the information above, it sounds like the teams themselves should be deploying the Service and the Application given the way the service is designed, using Data Sources to look up the relevant information needed to provision the Service if necessary.

Given it doesn't make sense to split this resource out in Terraform (since it'd make it harder for the majority of use-cases here) - whilst I'd like to thank @njucz for the contribution I'm going to close this PR for the moment.

Thanks!